### DISCUSSION TITLE
Thread DIDs

### DISCUSSION THREAD

**Discussion by sanderpick** [textileio/community#295] [Threads]
*An exploration of DIDs in a thread network.*

* [Motivation](#motivation)
* [What is a DID?](#what-is-a-did)
  * [Key points](#key-points)
* [DIDs in a Thread Network](#dids-in-a-thread-network)
  * [Peers](#peers-didkeyfoo)
  * [Users](#users-didkeyfoo-did3foo-didethrfoo-etc)
  * [Threads](#threads-didthreadid)
* [Thread DID Method Definition](#thread-did-method-definition)
  * [Method Scheme](#method-scheme)
  * [Method Operations](#method-operations)
    * [Create](#create)
    * [Read/Verify](#readverify)
    * [Update](#update)
    * [Deactivate](#deactivate)
    * [Security Requirements](#security-requirements)
    * [Privacy Requirements](#privacy-requirements)
* [Architecture](#architecture)
  * [Vanilla Network](#vanilla-network)
    * [Requirements](#requirements)
  * [Service-enabled Network](#service-enabled-network)
    * [Requirements](#requirements-1)
* [Examples](#examples)
  * [Dapps](#dapps)
  * [Smart Contracts](#smart-contracts)
  * [Exposing Web2 via Web3](#exposing-web2-via-web3)
* [References](#references)

# Motivation

[DIDs](https://w3c.github.io/did-core) (Decentralized Identifiers) enable cross-protocol/blockchain interactions by defining a common interface to retrieve and validate entities, like users and data.

Threads would benefit from a native DID in numerous ways:

1. Global namespaces. Thread IDs as unique identifiers to network-wide resources.
2. Ecosystem(s) interoperability. Integration with external DID methods available on systems such as Filecoin or Ethereum, as well as a range of Identity solutions (e.g., Ceramic/IDX).
3. Increased decentralization. Globally unique identifiers paves the way for distributed authorization to network resources, because access control mechanisms are globally defined.
4. Discovery mechanisms. Network peers can do work on each others' behalf since any peer is capable of discovering and resolving these namespaces.
5. Crypto native. Dapp developers can enable their users to directly leverage local crypto wallets to interact with thread-enabled apps.
6. Filecoin integration. Threads are append-only logs that can be anchored to Filecoin manually or at some frequency. Buckets are derivatives of threads that expose a mutable filesystem and pinning API that can be archived to Filecoin. In this sense, threads can become a Filecoin Layer 2 protocol.

# What is a DID?

A *DID* is a string identifier of a *subject*, controlled by a *controller*. For example, a DID might be used to [encode an Ethereum address](https://github.com/decentralized-identity/ethr-did-resolver#did-method) on the Ethereum Mainnet. For DIDs to be useful, they must be resolvable without reliance on a centralized network component. The figure below shows the interactions between DID components.

![DID Overview](https://github.com/textileio/community/assets/361000/7c306492-6c0c-4e12-b79e-9fc849e24f47)

_The basic components of DID architecture. Source: [https://w3c.github.io/did-core/#architecture-overview](https://w3c.github.io/did-core/#architecture-overview)_

The *Verifiable Data Registry* could be a blockchain or peer-to-peer network. The meaning of *verification* differs between DID implementations. Some provide a high level of verification through blockchain transactions, while others rely purely on the assumption that the majority of peers are *good actors,* e.g., IPID (an IPNS-based DID) or Ceramic.

## Key points

- A DID can identity any actor or structure in the network (the *subject*).
- A DID resolves to a *document*.
- A DID must be resolvable from anywhere in the network.
- A DID document describes the *subject*.
- A DID document is controlled by one or more *controllers*.

In short, DIDs point to relationships between the components of a decentralized network. What can this component do? Who/what can do it? Who/what dictates who can do it? Read the [full spec](https://w3c.github.io/did-core) for more info.

# DIDs in a Thread Network

A thread network has multiple actors and structures (subjects) that can be described by DID documents. Subjects include any of the following entities:

## Peers: `did:key:foo`

Peers (`go-threads` peers) can offer network services:

- Thread hosting
- Bucket hosting (IPFS, IPNS, pinning)
- Filecoin anchoring (by leveraging a local Powergate node)
- Database abstractions (see https://github.com/textileio/go-threads/discussions/485 for more)
- Service _hubs_ (see https://github.com/textileio/go-threads/discussions/484 for more)

A thread peer's DID is derived from its embedded LibP2P host key.

## Users: `did:key:foo`, `did:3:foo`, `did:ethr:foo`, etc.

A _User_ is any external identity that represents a network user, and that may interact with the network via a Peer. These may be identified by any verifiable DID.

## Threads: `did:thread:id`

Threads are collections of single-writer logs. Thread documents contain verification methods for all valid log signers. Other thread info like the *log head*, *log metadata*, and *thread encryption keys* are *not* stored in the DID document, since this information is only needed by peers that are sharing a thread, and can be more efficiently exchanged using the thread protocol vs. the global document registry. *Log addresses* are referenced as [service](https://www.w3.org/TR/did-core/#services) endpoints.

By identifying a thread as a global resource, any peer could determine the following from its DID:
1.  Who can write to the thread?
2.  Who can read from the thread?
3.  Where can the thread be bootstrapped from?
4.  Who controls (1), (2), and (3)?

Currently, this information is not globally available.

So, if the thread is the resource we're trying to identify, i.e., the thread _is the subject_ and the document itself is a [_representation of the subject_](https://www.w3.org/TR/did-core/#can-the-did-document-serve-as-a-representation-of-the-did-subject), then the DID [_subject is not the controller_](https://www.w3.org/TR/did-core/#set-2-the-did-subject-is-not-the-did-controller). The controller is one or more user DIDs, e.g., `did:key:foo`, `did:ethr:foo`, `did:3:foo`, etc.

The details of the document model aren't flushed out yet, but it might look something like the following:
```
{
  "@context": "https://www.w3.org/ns/did/v1",
  "id": "did:thread:id",
  "controller": "did:key:foo", // can modify this document
  "authentication": [
    {
      // can authenticate as did:thread:id
      "id": "did:key:foo#keys-1",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:foo",
      "publicKeyBase58": "..."
    },
    {
      // can authenticate as did:thread:id
      "id": "did:key:bar#keys-1",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:bar",
      "publicKeyBase58": "..."
    }
  ],
  "service": [
    {
      // can bootstrap did:thread:id
      "id": "did:thread:<id>#peer-id-1",
      "type": "ThreadService",
      "serviceEndpoint": "/p2p/peer-id-1",
      "serviceProtocol": "/thread/0.0.1"
    },
    {
      // can bootstrap did:thread:id
      "id": "did:thread:<id>#peer-id-2",
      "type": "ThreadService",
      "serviceEndpoint": "/p2p/peer-id-2",
      "serviceProtocol": "/thread/0.0.1"
    }
  ]
}
```

# Thread DID Method Definition

The crux of any DID implementation is defining it's [DID method](https://www.w3.org/TR/did-core/#methods). The DID spec defines a *method* as the following:

> DID methods provide the means to implement this specification on different verifiable data registries. New DID methods are defined in their own specifications, so that interoperability between different implementations of the same DID method is ensured.

In this section, we define a DID method specification for threads which is composed of a method *[scheme](https://www.w3.org/TR/did-core/#method-schemes)* and *[operations](https://www.w3.org/TR/did-core/#method-operations)*. [Operations](https://www.notion.so/Thread-DiD-2ad10636b05040ba81ae0674a4309d23) specify how a DID document is created, how to read/verify a document, as well as how a DID controller can update or even deactivate a DID document. Whereas the method [scheme](https://www.notion.so/Thread-DiD-2ad10636b05040ba81ae0674a4309d23) defines the structure of the DID implementation's string identifier(s).

## Method Scheme

```json
"did:<method>:<method-specific-id>"
```

_Generic DID structure._

A thread DID can simply prepend `did:thread:` to its [unique identifier](https://github.com/textileio/go-threads/blob/master/core/thread/id.go).

```json
"did:thread:<thread-id>"
```

_Thread DID structure._

```json
"did:thread:bafk6npbypzgshgb4b6hzjju7zdc3bfghx3ugacl6aie3xrfxcg4x6mfuhoe6iesr"
```

_A Thread DID._

## Method Operations

As mention perviously, DID methods define a set of operations that can be performed on/with the DID documents. DID implementations often use a smart contract on a blockchain like Ethereum to model the global data registry, and to implement these operations. For the implementation outlined here, we assume that documents are stored on the Filecoin blockchain as NFTs. Filecoin does not yet support NFTs, but there's an informal proposal to create a new *actor* type for them.

As a POC, we can develop a non-consensus driven global data registry using LibP2P PubSub along the lines of IPNS.

### Create

*To-do. DID documents are created by minting a new Filecoin NFT.*

### Read/Verify

The process of getting a DID subject's document from the verifiable data registry is called *resolution*. Any peer can resolve any document by querying the on-chain NFT representing the subject.

### Update

*To-do.*

### Deactivate

*To-do.*

## Security Requirements

*To-do.*

DID method specifications MUST include their own Security Considerations sections. This section MUST consider all the requirements mentioned in section 5 of [[RFC3552](https://www.w3.org/TR/did-core/#bib-rfc3552)] (page 27) for the DID operations defined in the specification.

## Privacy Requirements

*To-do.*

DID method specifications MUST include their own Privacy Considerations sections, if only to point to the DID spec, [§ 10. Privacy Considerations](https://www.w3.org/TR/did-core/#privacy-considerations).

# Architecture

The way in which a DID Method is leveraged is entirely up to the network itself. Here we outline some system requirements and walk through common network operations.

## Vanilla Network

First, let's consider a network of completely open peers. These can be local or remote. Open mean *no identity authorization.* Anyone can create/add threads.

### Requirements

- External identities leverage a peer (local or otherwise reachable by the user) to create threads.
- Once a thread is created, it can be considered globally available, i.e., thread peers can do work on behalf of other thread peers, and the globally unique identities mean any peer can do work on behalf of any other peer.
- Only peers that have been used to read/write to the thread will follow the thread.
- Any peer can be used to delete a thread.

![Vanilla Network](https://github.com/textileio/community/assets/361000/5d4351ff-a6b9-42e5-b34d-992452f9ed8b)

_A vanilla thread network (n=2) showing an external identity (A) creating a thread on one peer and writing to it from another peer._

## Service-enabled Network

A major promise of the decentralized web is putting ownership back into the hands of users by shifting publishing, discovery, and services from a few core providers, onto the "edge" and into the hands of users/creators. Services such as those traditionally found in centralized systems via open (or closed) application programing interfaces (APIs) are less common in p2p systems, however, a DID-driven thread network offers an exciting opportunity to expose p2p services as first-class components of a p2p system.

In this network, one of the peers has a *trusted relationship* with a service. Peer can advertise their services using the `"services"` DID field (read more about this [here](https://www.w3.org/TR/did-core/#services)). For example, let's consider a hypothetical web-hook service that allows users to add web-hooks to a thread. Every time the thread receives an update, the web-hook fires on the user-defined endpoint. The peer's DID document includes the service info:

```json
{
  "@context": "https://www.w3.org/ns/did/v1",
  "id": "did:key:peer-id",
  "authentication": [{
    "id": "did:key:peer-id#keys-1",
    "type": "Ed25519VerificationKey2018",
    "controller": "did:key:peer-id",
    "publicKeyBase58": "..."
  }],
  "services": [
    {
      "id": "did:key:peer-id#threads",
      "type": "ThreadService",
      "serviceEndpoint": "/p2p/peer-id",
      "serviceProtocol": "/thread/0.0.1"
    },
    {
      "id": "did:key:peer-id#webhooks",
      "type": "ThreadWebHookService",
      "serviceEndpoint": "/p2p/peer-id",
      "serviceProtocol": "/thread/0.0.1",
      "cost": {
        "hook": {
          "amount": "xx nanoFIL",
          "currency": "FIL"
        },
        "hit": {
          "amount": "xx nanoFIL",
          "currency": "FIL"
        }
      }
    }
  ]
}
```

### Requirements

- Services can define their own authorization patterns. Services may be free, have free quotas, only be open to some users, etc.
- A peer's services must be discoverable from their DID document.
- Services are self-describing via the DID service `"type"` , `"serviceEndpoint"`, and `"description"` fields. The community can maintain a list of available services by type.
- Peer/users can discover services on the network. Service discovery can be handled with LibP2P PubSub: Peers are used to request service types across the whole network. Matching host peers respond directly to the caller with a verifiable service description.

![Service-enabled Network](https://github.com/textileio/community/assets/361000/e9f4d06a-e0f7-47f2-bf5e-e754ef1caae7)

_A service-enabled thread network (n=2) with a hypothetical web-hook service. An external identity (A) has used the service to add a web-hook to a thread and then writes to the thread from another peer._

At a minimum, a peer will advertise it's own LibP2P thread API. Here are some examples of services a peer might offer:

- Buckets
    - Mutable filesystem API
    - Pinning API
- ThreadDB
- A `go-datastore` interface (key-value store)
- Web-hooks
- Filecoin
    - Thread anchoring
    - Bucket archiving
    - Deal retrievals
- MongoDB (`"serviceEndpoint": <connection_uri>`) *([see also Existing tools discussion](https://github.com/textileio/community/discussions/296))*
- PostgreSQL (`"serviceEndpoint": <connection_uri>`) *([see also Existing tools discussion](https://github.com/textileio/community/discussions/296))*
- Media encoding

The diagram below shows how buckets and Filecoin can be orchestrated as peer services.

![Bucket and Filecoin as Services](https://github.com/textileio/community/assets/361000/9ffd04b7-1ff3-4b45-8496-2e56c3c472b2)

_Buckets and Filecoin can be considered network services._

# Examples

## Dapps

[Slate](https://slate.host/) is a good example of an application that could be developed as a pure dapp on the thread network. Many other apps can leverage these same patterns: Slack, Notion, etc.

- User's become a Filecoin wallet/address specified with a simple key DID: `"did:fil:<address>"`.
- Slates are buckets in the user's slates thread. There could be a need for more complexity here depending on the needed metadata, but this model is enough to illustrate the idea.
- A user's slate thread and slates (buckets) are registered DID subjects controlled by their wallet/address DID (via the `"controller"` field).
- Users can collaborate on slates by updating the DID document for the thread to include a verification method for *their* identity, allowing them to read/write from/to the thread.
- Users can fund Filecoin archives of their slates directly from their wallet/address.
- Chat can be added in the same way as a collaborative slate, but instead of media, the content is text (or text and media). The difference is mostly in the UI.
- **The application can be run entirely client side against a remote or local thread peer.**

## Smart Contracts

Smart contracts on any blockchain can access IPFS and Filecoin storage without the need for a centralized pinning service. This is possible because users (or client-side peers) are able to query the registry without depending on a *specific peer* or set of peers. Client-side peers may be lightweight thread peers with read-only capabilities, or with the ability to query but not respond to registry events. All of this enables the use of a single token to pin content against a given peer (maybe the hub, or some bucket pinning API) without the need for a secret. This is because the receiver can *validate* that the caller is authorized to do so *by looking at the thread DID doc*.

## Exposing Web2 via Web3

In tandem with an additional threads effort to [wrap existing tools](https://github.com/textileio/community/discussions/296), a thread network can enable existing web2 services, APIs, and applications to read and write data using existing tools like MongoDB, PostgreSQL, Redis, etc. This opens Filecoin to a massive untapped source of data. But the thread network also allows service providers to take this one step further.

In addition to exposing more "traditional" web2 APIs (REST, gRPC, etc) as services, it is possible to expose these APIs over p2p protocols. Projects such as [libp2p-http](https://github.com/libp2p/go-libp2p-http), allow peers to serve HTTP endpoints and make HTTP requests through libp2p using Go's standard "http" and "net" stack. This provides a simple on-ramp for web2 developers to expose their APIs over the threads p2p network.

For example, a peer may serve a RESTful pinning API that they expose via a p2p service. Any service that produces content-addressable outputs can be easily federated via the registry, and any "location-specific" APIs (e.g., direct database access) can benefit from libp2p-based routing — taking advantage of features like multi-routes, NAT transversal and stream multiplexing over a single connection. Additionally, these APIs can now take advantage of the authorization patterns of the threads network "for free", providing direct benefits to service providers, and opening the door to crypto-native payments for off-chain services.

# References

- [https://github.com/WebOfTrustInfo/rwot5-boston/blob/master/topics-and-advance-readings/did-primer.md](https://github.com/WebOfTrustInfo/rwot5-boston/blob/master/topics-and-advance-readings/did-primer.md)
- [https://github.com/WebOfTrustInfo/rwot5-boston/blob/master/topics-and-advance-readings/verifiable-claims-primer.md](https://github.com/WebOfTrustInfo/rwot5-boston/blob/master/topics-and-advance-readings/verifiable-claims-primer.md)
- [https://www.w3.org/TR/did-core](https://www.w3.org/TR/did-core)
- [https://w3c.github.io/did-spec-registries](https://w3c.github.io/did-spec-registries/)
- [https://tsmatz.wordpress.com/2020/06/25/what-is-verifiable-credentials](https://tsmatz.wordpress.com/2020/06/25/what-is-verifiable-credentials/)
- [https://github.com/ockam-network/did-method-spec](https://github.com/ockam-network/did-method-spec)
- [https://github.com/ontio/ontology-go-sdk](https://github.com/ontio/ontology-go-sdk)
- [https://excalidraw.com/#json=6542356975190016,FyU14p_VS6LPk49GU9_UPg](https://excalidraw.com/#json=6542356975190016,FyU14p_VS6LPk49GU9_UPg)

**Comment by sanderpick** [comment]
Howdy all! Here's a related discussion about service hubs: https://github.com/textileio/go-threads/discussions/484

**Comment by carsonfarmer** [comment]
And here's a related discussion about integrating existing tools and services (see also the services discussion in _this_ discussion): https://github.com/textileio/go-threads/discussions/485

**Comment by andrewxhill** [comment]
Another related discussion is the [Bucket Pinning API for IPFS & Filecoin](https://github.com/textileio/textile/discussions/499). It will likely be one of the first APIs to support some of the above methods, so take a look over there if interested.

**Comment by oed** [comment]
DIDs for peers, users, and projects makes a ton of sense and I'm amped to see support for this in the threads protocol!

I don't exactly follow what the benefit of using DIDs for the threads themselves are? Usually a DID and it's corresponding document is a way to bootstrap trust from an identifier to a set of public keys (which might change over time) that the subject controls. Threads don't really have control any private keys so the main value prop for using DIDs is unclear to me.

**Reply by sanderpick** [reply]
Hey @oed, thank you for diving in here! Maybe you can help me figure out if we're off base.

The [ethr-did-registry](https://github.com/uport-project/ethr-did-registry) says, "A DID is an Identifier that allows you to lookup a DID document that can be used to authenticate you and messages created by you."

That seems to be a narrow interpretation of the spec. The section [D.1 What types of resources can a DID identify?](https://www.w3.org/TR/did-core/#what-types-of-resources-can-a-did-identify) states the following:

> Since a DID is a specific type of URI (Uniform Resource Identifier), the answer to this question is provided by section 1.1 of the URI specification [RFC3986]:

>> This specification does not limit the scope of what might be a resource; rather, the term "resource" is used in a general sense for whatever might be identified by a URI. Familiar examples include an electronic document, an image, a source of information with a consistent purpose (e.g., "today's weather report for Los Angeles"), a service (e.g., an HTTP-to-SMS gateway), and a collection of other resources. A resource is not necessarily accessible via the Internet; e.g., human beings, corporations, and bound books in a library can also be resources. Likewise, abstract concepts can be resources, such as the operators and operands of a mathematical equation, the types of a relationship (e.g., "parent" or "employee"), or numeric values (e.g., zero, one, and infinity).

By identifying a thread as a global resource, any peer could determine the following from its DID:
1.  Who can write to the thread?
2.  Who can read from the thread?
3.  Where can the thread be bootstrapped from?
4.  Who controls (1), (2), and (3)?

Currently, this information is not globally available.

So, if the thread is the resource we're trying to identify, i.e., the thread _is the subject_ and the document itself is a [_representation of the subject_](https://www.w3.org/TR/did-core/#can-the-did-document-serve-as-a-representation-of-the-did-subject), then the DID [_subject is not the controller_](https://www.w3.org/TR/did-core/#set-2-the-did-subject-is-not-the-did-controller). The controller is one or more user DIDs, e.g., `did:key:foo`, `did:ethr:foo`, `did:3:foo`, etc.

The details of the document model aren't flushed out yet, but it might look something like the following:
```
{
  "@context": "https://www.w3.org/ns/did/v1",
  "id": "did:thread:id",
  "controller": "did:key:foo", // can modify this document
  "authentication": [
    {
      // can authenticate as did:thread:id
      "id": "did:key:foo#keys-1",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:foo",
      "publicKeyBase58": "..."
    },
    {
      // can authenticate as did:thread:id
      "id": "did:key:bar#keys-1",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:bar",
      "publicKeyBase58": "..."
    }
  ],
  "service": [
    {
      // can bootstrap did:thread:id
      "id": "did:thread:<id>#peer-id-1",
      "type": "ThreadService",
      "serviceEndpoint": "/p2p/peer-id-1",
      "serviceProtocol": "/thread/0.0.1"
    },
    {
      // can bootstrap did:thread:id
      "id": "did:thread:<id>#peer-id-2",
      "type": "ThreadService",
      "serviceEndpoint": "/p2p/peer-id-2",
      "serviceProtocol": "/thread/0.0.1"
    }
  ]
}
```

**Reply by oed** [reply]
Thanks for clarifying @sanderpick. I had not really thought about this type of use case for DIDs before so it threw me off a bit, but seems like it makes sense 👍 

One thing to look out for: 
In the `authentication` section the public key entries require that the `id` property have a "fragment" i.e. `#key-name`. I the DID that this refers to rotates keys it's likely that this key is revoked in favor of a new key, which would revoke their access from the thread. Unfortunately I think it's impossible to specify the `id` without the fragment right now (and this might not be what is really desirable here either, not sure). These are hard problems in general, so happy to help figuring these things out :)

**Comment by Schwartz10** [comment]
Hey @sanderpick this is exciting and really well articulated. I'm still trying to wrap my head around all of the moving parts, but some initial questions as they relate to daemon.land (The project my team and I are working on, which is working on Textile integrations with Ceramic/IDX):

> By identifying a thread as a global resource, any peer could determine the following from its DID: 
    Who can write to the thread?
    Who can read from the thread?
    Where can the thread be bootstrapped from?
    Who controls (1), (2), and (3)?

A couple questions here, referencing the data structure you pasted:

```js
"authentication": [
    {
      // can authenticate as did:thread:id
      "id": "did:key:foo#keys-1",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:foo",
      "publicKeyBase58": "..."
    },
    {
      // can authenticate as did:thread:id
      "id": "did:key:bar#keys-1",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:bar",
      "publicKeyBase58": "..."
    }
  ],
```
1. To make sure I'm understanding correctly, the answers to the access control questions (1 and 2) above can be found from the `authentication` key in the DID Document right? 
2. Given this document, how do you actually interpret read and write access? Do I have to know anything about Textile in order to make that interpretation? Is it possible to grant one (read OR write) without the other? How flexible can we make the permission systems?
3.  Can `did:key:bar` or `did:key:foo` be any DID or does that _have_ to be a thread peer in order to integrate natively into the threads network?

> Other thread info like the log head, log metadata, and thread encryption keys are not stored in the DID document

I generally understand this (esp when it comes to frequently changing information like log heads), but just want to bring up a couple counter arguments in case they're useful (not sure where I personally stand on this yet). First,  in _some_ circumstances storing some metadata in the DID document can help promote interoperability between things. For example, storing the thread's schema would help other developers reuse threads when applicable, which is important in promoting application level interop. Second, storing encrypted encryption keys on the Document itself could enable certain "thread services" to roll their own access control mechanisms on top of threads, so you guys wouldn't have to care (as much) about supporting certain access mechanisms. You could encrypt to the thread DID, so only authenticated controllers could decrypt certain keys? (need to think about that more)

> The community can maintain a list of available services by type.

I think this is a fantastic idea, and something we've thought of for maintaining a list of popularly used schemas as well. Is this something Textile would promote and manage? Where would that list live? How would it get updated? How would users and developers know which to choose from and which to trust?

Generally after reading this, I think thread DIDs sound like a great idea to move the system forward and enable more p2p interactions. However, still the big Q on my end is about more flexible thread and bucket access control mechanisms, and being able to dynamically alter them without causing massively expensive operations.

**Reply by sanderpick** [reply]
Hey @Schwartz10, thanks for chiming in! 

> To make sure I'm understanding correctly, the answers to the access control questions (1 and 2) above can be found from the authentication key in the DID Document right?

Yep, that's the idea anyway

> Given this document, how do you actually interpret read and write access? Do I have to know anything about Textile in order to make that interpretation? Is it possible to grant one (read OR write) without the other? How flexible can we make the permission systems?

Yep, but I'm not sure about the best way to express those rules. I've been experimenting with a couple ideas, but the DID spec leaves ample room for interpretation of things like [capability-invocation](https://www.w3.org/TR/did-core/#capability-invocation).

Do you have any specific requirements?

> Can did:key:bar or did:key:foo be any DID or does that have to be a thread peer in order to integrate natively into the threads network?

The goal is for that to be _any_ DID, but in practice we'll initially support `did:key`, `did:3`, and `did:ethr`. @carsonfarmer has started working on Go resolvers for those [here](https://github.com/textileio/go-did-resolver).

> storing the thread's schema would help other developers reuse threads when applicable, which is important in promoting application level interop. 

To clarify, a thread DID describes the lower level thread concept: a collection of logs. The DB layer that deals in schemas becomes a service: ThreadDB. That said, the idea of service metadata is really nice. In the current case of buckets, that metadata live _in the thread_, but that may be cumbersome for other services. Need to think more about it.

> storing encrypted encryption keys on the Document itself could enable certain "thread services" to roll their own access control mechanisms on top of threads, so you guys wouldn't have to care (as much) about supporting certain access mechanisms. You could encrypt to the thread DID, so only authenticated controllers could decrypt certain keys? (need to think about that more)

Nice, offloading things like access control sounds great. Storing encrypted data in a DID document sounds dangerous and [seems to be discouraged](https://www.w3.org/TR/did-core/#encrypted-data-in-did-documents), but maybe that should be an application level concern.

**Comment by christroutner** [comment]
@andrewxhill passed me the link to this Issue when I started telling him about the [ipfs-coord](https://github.com/christroutner/ipfs-coord) npm library I've been working on. There is large overlap between that idea and this one. I've been searching for other people interested in this topic. I've even been posting on the [IPFS discussion forum](https://discuss.ipfs.io/t/ipfs-replacement-for-rest-api/10054).

[This blog post](https://troutsblog.com/blog/ipfs-api/) lays out my ideas, which could be considered a slightly different tweak to the ideas presented here. [This YouTube video presentation](https://youtu.be/VVc0VbOD4co) present my early conceptualization of research into the idea.

I love the idea of the DID and using a self-describing JSON document. This just happens to be the part of the problem I've been looking for a better solution to. This proposal seems like a fine way to do it. The DID document presented here could take the place of [the schema](https://github.com/christroutner/ipfs-coord/blob/master/lib/schema.js) library I'm using now.

> In addition to exposing more "traditional" web2 APIs (REST, gRPC, etc) as services, it is possible to expose these APIs over p2p protocols.

I'd love to participate in nailing down specifics around an RPC specification. I'm just about to wrap up e2e encryption. I've been eyeballing this [jsonrpc-lite](https://www.npmjs.com/package/jsonrpc-lite) npm library as way to bootstrap a very simple RPC system. This library would allow easy retrofitting of REST APIs to provide their same web services over IPFS.

I've got an early prototype demo using the ipfs-coord library at [chat.FullStack.cash](https://chat.fullstack.cash). It's a real-time chat app. I'm hoping to find time to record a presentation on my progress and the inner workings of the demo and ipfs-coord library, once I finish getting e2e encryption and private chat working in the demo.

**Reply by carsonfarmer** [reply]
Cool stuff here @christroutner, happy to have you as part of the discussion, especially around the RPC stuff. I think we'll likely do some similar "discussion" style posts like this for things like this. I also have some work on generic rpc over libp2p streams, as well as http, and json rpc. I haven't "released" these 2nd two yet, but I'll aim to do that this week in case they are of interest.

**Reply by sanderpick** [reply]
Thanks for sharing, @christroutner. Very interesting! We'd love you help in figuring this all out. 

> I'd love to participate in nailing down specifics around an RPC specification.

As a POC, I've been using gRPC over libp2p using https://github.com/libp2p/go-libp2p-core/blob/master/mux/mux.go. Perhaps we can use a similar approach for a JSON RPC API that is more JS friendly. In a past project, we wrote a custom RPC mechanism on raw libp2p, and it was pretty awful to deal with.

**Reply by carsonfarmer** [reply]
Yeh nice. I'm defs a fan of JSON RPC, though it is a bit limiting in terms of the types of calls you can make, so something more generic, but still nicer for browsers would be really nice. This is a good start (https://github.com/carsonfarmer/libp2p-rpc), but it uses `msgpack5`. Not a problem per se, but there might be nicer wire formats like cbor that we could standardize around. Additionally, the above setup that @sanderpick is talking about actually supports pretty much any web-protocols we might want to support. So there's nothing stopping us from exposing JSON RPC, HTTP (REST) APIs, gRPC, and more. Once you're dealing with low level sockets, everything is pretty much wide open again :)

As a point of comparison, Ethereum nodes provide JSON RPC (among other) endpoints: https://eth.wiki/json-rpc/API, so there is some standardization there.

**Reply by sanderpick** [reply]
The crux is deciding how to type the request/response data. True, we can expose any number of endpoints, but I think having a single way to represent data is important. IPFS uses protos and CBOR. Perhaps supporting both wouldn't be too much of a pain.

**Reply by christroutner** [reply]
I'm glad you're receptive to the work I've been doing!

Picking a good RPC solution has a high risk of bike-shedding. There are many solutions to the problem, and everyone is going to gravitate toward their favorite. The determining factors in my mind is:
- Simplicity: The solutions should be simple, and strike a good balance between network efficiency and easy debugging. This is why I'm a big fan of the JSON RPC. It's efficient 'enough', and it's human-readable enough to make debugging really easy.
- Standards: This is not a new problem. It's been solved many times. Using a standard will allow the overarching idea to tap into network effects by adopting a common standard.

**Comment by sanderpick** [comment]
Ping @dgtony and @requilence. We'd love some feedback from the Anytype team. I want to make sure we're solving something for your use cases.

**Comment by christroutner** [comment]
I wanted to follow up on my previous comment. I've just published a couple demos to YouTube, which are linked in [this blog post](https://troutsblog.com/blog/chat-fullstack-cash). I link to this thead and give a shout-out to the work done here. This represents the latest in my ongoing research.

I'm all ears if you have any feedback.

**Reply by sanderpick** [reply]
Hey @christroutner, great videos. Thanks for sharing. It really does sound like we're well aligned here. We haven't done anything here on the JS end yet. In addition to the RPC design, @carsonfarmer there might be an opportunity here to collab on the JS libp2p host management: Circuit relays, etc.

**Reply by christroutner** [reply]
I appreciate the feedback!

I've also been brushing up on the DID specification, as well as JSON-LD. 

I'd love to see Textile build out their own DID implementation. That's certainly beyond the scope of what I want to do. I've decided to move forward with JSON-LD, focusing on the schema.org syntax. That will dovetail nicely with any DID implementation that might come down the road.

I'm going to continue to improve [chat.fullstack.cash](https://chat.fullstack.cash), and it will eventually be forked into two apps: a database of deep-web websites, and a decentralized exchange (DEX) for cross-blockchain sale of cryptocurrencies. See this [Twitter thread](https://twitter.com/christroutner/status/1375998642352316421) for more info.

I'll continue to make efforts to stay in the loop, and look for opportunities to incorporate any work from Textile. I need to do a little deeper dive into the concepts behind Threads, and how they differ from OribitDB.

**Comment by christroutner** [comment]
I came across this [jsonld](https://github.com/digitalbazaar/jsonld.js) npm library an it lead me down the rabbit hole of the JSON-LD specification. I think this knowledge is complimentary to the idea of a DID and developing an API for decentralized services. This page has a lot of short introductory YouTube videos to the core concepts around JSON-LD:

- https://json-ld.org/learn.html
